### ISSUE TITLE
verification pass and well-formedness check

### ISSUE THREAD

**Issue by vzaliva** [vellvm/vellvm#142]
There is a notion of [well-formed IR](https://llvm.org/docs/LangRef.html#well-formedness) which is ensured by special LLVM [verification pass](https://llvm.org/docs/Passes.html#verify-module-verifier). Additionally, LLVM documentation states:

> "Note that llvm-as verifies its input before emitting bitcode, and also that malformed bitcode is likely to make LLVM crash. All language front-ends are therefore encouraged to verify their output before performing optimizing transformations."

When synthesizing IR programs using Vellvm, it may be useful to guarantee that they are well-formed. For example, this could be ensured by defining the `isWellFormed` predicate which guarantees the following properties:

1. Both of a binary operator’s parameters are of the same type.
2. Verify that the indices of mem access instructions match other operands.
3. Verify that arithmetic and other things are only performed on first-class types. Verify that shifts and logicals only happen on integrals f.e.
4. All of the constants in a switch statement are of the correct type.
5. The code is in valid SSA form.
6. It is illegal to put a label into any other type (like a structure) or to return one.
7. Only phi nodes can be self-referential: `%x = add i32 %x, %x` is invalid.
8. PHI nodes must have an entry for each predecessor, with no extras.
9. PHI nodes must be the first thing in a basic block, all grouped together.
10. PHI nodes must have at least one entry.
11. All basic blocks should only end with terminator insts, not contain them.
12. The entry node to a function must not have predecessors.
13. All Instructions must be embedded into a basic block.
14. Functions cannot take a void-typed parameter.
15. Verify that a function’s argument list agrees with its declared type.
16. It is illegal to specify a name for a void value.
17. It is illegal to have an internal global value with no initializer.
18. It is illegal to have a ret instruction that returns a value that does not agree with the function return value type.
19. Function call argument types match the function prototype.
20. All other things that are tested by asserts spread about the code.

**Comment by Zdancewic** [comment]
I agree that it would be great to create a formalized well-formedness checker for LLVM IR as part of Vellvm, especially if it was extractable / executable.  Doing so could help debug certain program transformations and compilers that target LLVM IR.  

However, that static semantics is largely orthogonal to the behavioral semantics of an LLVM IR program -- it would be hard to take advantage of well-formedness when implementing the denotational semantics, for instance.  The dynamic semantics can assume that the LLVM IR input it well formed and should be permitted to exhibit any behavior for non-well-formed programs.

**Comment by vzaliva** [comment]
Agreed that semantics does not have to directly depend on well-formedness. However, compiler frontends need to produce well-formed IR programs to be accepted by the LLVM backend. Thus compiler writers using Vellvm may want to ensure (prove) this property in addition to semantic preservation or other properties they prove for their compiler.

For example, consider a program that violates the SSA form by doing a duplicate identical assignment of the same variable. Semantics may allow this, and semantically the second duplicate assignment will have no effect. However, being non-well-formed it will not be accepted by the IR backend compiler.

**Comment by YaZko** [comment]
I think Steve was saying that a well-formedness _checker_ would be a cool but unneeded luxury. A well-formedness propositional predicate however will certainly come in handy as soon as we prove optimizations whose correctness will rely on this predicate holding.

And similarly I agree that verified front end should establish the predicate over the generated code (although they should be able to satisfy the SSA part trivially by putting everything in memory and relying on the `mem2reg` pass to do the job I believe).

**Comment by YaZko** [comment]
Depending on what we want to do, #127 could be a particular case of such a broad well-formedness predicate.

The shape and content of the relevant predicate, or predicates, is still for now unclear and will be refined by need I believe. I'll start working next month with a student on the design of code generators enforcing static guarantees. It will be geared mainly toward scoping issues, but might also cover part of these questions.

I'll leave this Issue open and close #127 in the mean time if that's alright, don't hesitate to protest otherwise :)

**Comment by YaZko** [comment]
While this remark is still relevant, we are moving it to internal notes to keep the issues themselves more focused.
